File: go/ontology/classify_test.go
// SPDX-License-Identifier: Apache-2.0
package ontology_test

import (
	"context"
	"testing"

	"github.com/jeffs-brain/memory/go/llm"
	"github.com/jeffs-brain/memory/go/ontology"
)

// fakeProvider is a mock LLM provider for testing classification.
type fakeProvider struct {
	response string
	err      error
}

func (f *fakeProvider) Complete(_ context.Context, _ llm.CompleteRequest) (llm.CompleteResponse, error) {
	if f.err != nil {
		return llm.CompleteResponse{}, f.err
	}
	return llm.CompleteResponse{Text: f.response}, nil
}

func (f *fakeProvider) CompleteStream(_ context.Context, _ llm.CompleteRequest) (<-chan llm.StreamChunk, error) {
	return nil, nil
}

func (f *fakeProvider) Close() error { return nil }

func TestIsJsonDocument_ValidObject(t *testing.T) {
	t.Parallel()
	content := `{"rules": [{"name": "discount"}, {"name": "tax"}]}`
	if !ontology.IsJsonDocument(content) {
		t.Fatal("expected true for valid JSON object")
	}
}

func TestIsJsonDocument_ValidArray(t *testing.T) {
	t.Parallel()
	content := `[{"name": "Alice"}, {"name": "Bob"}]`
	if !ontology.IsJsonDocument(content) {
		t.Fatal("expected true for valid JSON array")
	}
}

func TestIsJsonDocument_InvalidJson(t *testing.T) {
	t.Parallel()
	content := "This is just some regular text content."
	if ontology.IsJsonDocument(content) {
		t.Fatal("expected false for prose text")
	}
}

func TestIsJsonDocument_TrivialJson(t *testing.T) {
	t.Parallel()
	// Bare primitives are not business-relevant JSON
	if ontology.IsJsonDocument("42") {
		t.Fatal("expected false for bare number")
	}
	if ontology.IsJsonDocument(`"hello"`) {
		t.Fatal("expected false for bare string")
	}
}

func TestIsTabularDocument_Csv(t *testing.T) {
	t.Parallel()
	content := "name,age,city\nAlice,30,NYC\nBob,25,LA\nCharlie,35,SF\n"
	if !ontology.IsTabularDocument(content) {
		t.Fatal("expected true for CSV content")
	}
}

func TestIsTabularDocument_Pipe(t *testing.T) {
	t.Parallel()
	content := "name|age|city\nAlice|30|NYC\nBob|25|LA\nCharlie|35|SF\n"
	if !ontology.IsTabularDocument(content) {
		t.Fatal("expected true for pipe-delimited content")
	}
}

func TestIsTabularDocument_Prose(t *testing.T) {
	t.Parallel()
	content := "This is a paragraph of text.\nAnother paragraph follows.\nNo delimiters here.\n"
	if ontology.IsTabularDocument(content) {
		t.Fatal("expected false for prose text")
	}
}

func TestClassify_JsonDetected(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	c := ontology.NewClassifier(nil)

	result, err := c.Classify(ctx, `{"customers": [{"id": 1}]}`, "data.json")
	if err != nil {
		t.Fatalf("Classify: %v", err)
	}
	if result.Class != ontology.DocumentClassStructured {
		t.Fatalf("expected structured, got %q", result.Class)
	}
	if !result.IsStructured {
		t.Fatal("expected IsStructured to be true")
	}
	if result.Confidence < 0.9 {
		t.Fatalf("expected high confidence, got %f", result.Confidence)
	}
}

func TestClassify_TabularDetected(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	c := ontology.NewClassifier(nil)

	content := "name,age,city\nAlice,30,NYC\nBob,25,LA\nCharlie,35,SF\n"
	result, err := c.Classify(ctx, content, "contacts.csv")
	if err != nil {
		t.Fatalf("Classify: %v", err)
	}
	if result.Class != ontology.DocumentClassTabular {
		t.Fatalf("expected tabular, got %q", result.Class)
	}
	if !result.IsStructured {
		t.Fatal("expected IsStructured to be true for tabular")
	}
}

func TestClassify_FallbackToLLM(t *testing.T) {
	t.Parallel()
	ctx := context.Background()

	provider := &fakeProvider{
		response: `{"category": "entity", "confidence": 0.85, "reasoning": "contains customer data"}`,
	}
	c := ontology.NewClassifier(provider)

	content := "The customer approval process requires manager sign-off for orders above $10,000."
	result, err := c.Classify(ctx, content, "approval-rules.md")
	if err != nil {
		t.Fatalf("Classify: %v", err)
	}
	if result.Class != ontology.DocumentClassUnstructured {
		t.Fatalf("expected unstructured, got %q", result.Class)
	}
	if result.Confidence <= 0 {
		t.Fatal("expected non-zero confidence from LLM")
	}
}

func TestClassify_NoProvider(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	c := ontology.NewClassifier(nil)

	content := "The customer approval process requires manager sign-off."
	result, err := c.Classify(ctx, content, "doc.md")
	if err != nil {
		t.Fatalf("Classify: %v", err)
	}
	if result.Class != ontology.DocumentClassUnstructured {
		t.Fatalf("expected unstructured, got %q", result.Class)
	}
	if result.Category != "general" {
		t.Fatalf("expected general category without provider, got %q", result.Category)
	}
}

func TestClassify_JsonWithCustomerKeywords(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	c := ontology.NewClassifier(nil)

	content := `{"customer_id": 123, "customer_name": "Acme Corp", "account_type": "enterprise"}`
	result, err := c.Classify(ctx, content, "data.json")
	if err != nil {
		t.Fatalf("Classify: %v", err)
	}
	if result.Category != "customer" {
		t.Fatalf("expected customer category, got %q", result.Category)
	}
}

func TestDetermineCategory_WithOntology(t *testing.T) {
	t.Parallel()
	ontologyData := &ontology.ResolvedOntology{
		NodeTypes: []ontology.ResolvedType{
			{TypeDefinition: ontology.TypeDefinition{
				Type: "entity.customer", Label: "Customer",
				Description: "Customer entity for customer management",
				CreatedAt: "2026-01-01T00:00:00Z", Status: "active",
			}},
		},
		BusinessCategories: []string{"customer", "order", "general"},
	}

	content := "The customer submitted a new order for processing."
	category := ontology.DetermineCategory(content, ontologyData)
	// Should find "customer" in the content matching entity.customer
	if category == "" {
		t.Fatal("expected non-empty category")
	}
}

func TestDetermineCategory_NilOntology(t *testing.T) {
	t.Parallel()
	category := ontology.DetermineCategory("some content", nil)
	if category != "general" {
		t.Fatalf("expected general for nil ontology, got %q", category)
	}
}

func TestCategoryWinnerThreshold(t *testing.T) {
	t.Parallel()
	// Test that the threshold constant matches the intelligence service
	if ontology.CategoryWinnerThreshold != 0.4 {
		t.Fatalf("expected 0.4, got %f", ontology.CategoryWinnerThreshold)
	}
}

File: sdks/ts/memory/src/ontology/classify.test.ts
// SPDX-License-Identifier: Apache-2.0

import { describe, expect, it } from 'vitest'
import type { Provider, CompletionResponse } from '../llm/types.js'
import type { ResolvedOntology, ResolvedType } from './store.js'
import {
  Classifier,
  isJsonDocument,
  isTabularDocument,
  determineCategory,
  CATEGORY_WINNER_THRESHOLD,
} from './classify.js'

function fakeProvider(response: string): Provider {
  return {
    name: () => 'fake',
    modelName: () => 'fake-model',
    stream: () => {
      throw new Error('not implemented')
    },
    complete: async (): Promise<CompletionResponse> => ({
      content: response,
      toolCalls: [],
      usage: { inputTokens: 0, outputTokens: 0 },
      stopReason: 'end_turn',
    }),
    supportsStructuredDecoding: () => false,
    structured: async () => response,
  }
}

describe('isJsonDocument', () => {
  it('returns true for valid JSON object', () => {
    expect(isJsonDocument('{"rules": [{"name": "discount"}]}')).toBe(true)
  })

  it('returns true for valid JSON array', () => {
    expect(isJsonDocument('[{"name": "Alice"}, {"name": "Bob"}]')).toBe(true)
  })

  it('returns false for prose text', () => {
    expect(isJsonDocument('This is just some regular text content.')).toBe(false)
  })

  it('returns false for bare primitives', () => {
    expect(isJsonDocument('42')).toBe(false)
    expect(isJsonDocument('"hello"')).toBe(false)
  })

  it('returns false for empty string', () => {
    expect(isJsonDocument('')).toBe(false)
  })

  it('returns false for primitive arrays', () => {
    expect(isJsonDocument('[1, 2, 3]')).toBe(false)
    expect(isJsonDocument('["a", "b", "c"]')).toBe(false)
  })

  it('returns false for empty structures', () => {
    expect(isJsonDocument('{}')).toBe(false)
    expect(isJsonDocument('[]')).toBe(false)
  })
})

describe('isTabularDocument', () => {
  it('returns true for CSV content', () => {
    const content = 'name,age,city\nAlice,30,NYC\nBob,25,LA\nCharlie,35,SF\n'
    expect(isTabularDocument(content)).toBe(true)
  })

  it('returns true for pipe-delimited content', () => {
    const content = 'name|age|city\nAlice|30|NYC\nBob|25|LA\nCharlie|35|SF\n'
    expect(isTabularDocument(content)).toBe(true)
  })

  it('returns false for prose text', () => {
    const content = 'This is a paragraph.\nAnother paragraph follows.\nNo delimiters here.\n'
    expect(isTabularDocument(content)).toBe(false)
  })
})

describe('Classifier', () => {
  describe('classify', () => {
    it('detects JSON as structured', async () => {
      const c = new Classifier({})
      const result = await c.classify('{"customers": [{"id": 1}]}', 'data.json')
      expect(result.class).toBe('structured')
      expect(result.isStructured).toBe(true)
      expect(result.confidence).toBeGreaterThanOrEqual(0.9)
    })

    it('detects CSV as tabular', async () => {
      const c = new Classifier({})
      const content = 'name,age,city\nAlice,30,NYC\nBob,25,LA\nCharlie,35,SF\n'
      const result = await c.classify(content, 'contacts.csv')
      expect(result.class).toBe('tabular')
      expect(result.isStructured).toBe(true)
    })

    it('falls back to LLM for unstructured content', async () => {
      const provider = fakeProvider('{"category": "entity", "confidence": 0.85, "reasoning": "customer data"}')
      const c = new Classifier({ provider })
      const content = 'The customer approval process requires manager sign-off for orders above $10,000.'
      const result = await c.classify(content, 'approval-rules.md')
      expect(result.class).toBe('unstructured')
      expect(result.confidence).toBeGreaterThan(0)
    })

    it('returns default without provider', async () => {
      const c = new Classifier({})
      const content = 'The customer approval process requires manager sign-off.'
      const result = await c.classify(content, 'doc.md')
      expect(result.class).toBe('unstructured')
      expect(result.category).toBe('general')
    })

    it('infers customer category from JSON keywords', async () => {
      const c = new Classifier({})
      const content = '{"customer_id": 123, "customer_name": "Acme Corp", "account_type": "enterprise"}'
      const result = await c.classify(content, 'data.json')
      expect(result.category).toBe('customer')
    })
  })
})

describe('determineCategory', () => {
  const sampleOntology: ResolvedOntology = {
    nodeTypes: [
      {
        type: 'entity.customer',
        label: 'Customer (Entity)',
        description: 'Customer entity for customer management',
        createdAt: '2026-01-01T00:00:00Z',
        status: 'active',
        scope: 'built-in',
      } as ResolvedType,
    ],
    edgeTypes: [],
    businessCategories: ['customer', 'order', 'general'],
  }

  it('returns specific category from ontology analysis', () => {
    const content = 'The customer submitted a new order for processing.'
    const category = determineCategory(content, sampleOntology)
    // Content contains "customer" which matches entity.customer whose
    // description contains "customer" (a business category).
    expect(category).toBe('customer')
  })

  it('returns general for undefined ontology', () => {
    expect(determineCategory('some content', undefined)).toBe('general')
  })
})

describe('LLM category mapping', () => {
  it('preserves LLM process category instead of mapping to general', async () => {
    const provider = fakeProvider('{"category": "process", "confidence": 0.9, "reasoning": "describes a workflow"}')
    const c = new Classifier({ provider })
    const content = 'The approval workflow requires two levels of sign-off before any purchase order is released.'
    const result = await c.classify(content, 'workflow.md')
    expect(result.category).toBe('process')
  })

  it('preserves LLM entity category', async () => {
    const provider = fakeProvider('{"category": "entity", "confidence": 0.85, "reasoning": "contains customer data"}')
    const c = new Classifier({ provider })
    const content = 'The customer approval process requires manager sign-off for orders above $10,000.'
    const result = await c.classify(content, 'approval-rules.md')
    expect(result.category).toBe('entity')
  })
})

describe('constants', () => {
  it('CATEGORY_WINNER_THRESHOLD matches intelligence service', () => {
    expect(CATEGORY_WINNER_THRESHOLD).toBe(0.4)
  })
})
